libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -10,157 +10,269 @@
-                var newValue = change.Value;
-                StoreItemEntity entity = new StoreItemEntity(entityKey, newValue);
-                if (entity.ETag == null || entity.ETag == "*")
-                {
-                    var result = await this.Table.ExecuteAsync(TableOperation.InsertOrReplace(entity)).ConfigureAwait(false);
-                }
-                else if (entity.ETag.Length > 0)
-                {
-                    var result = await this.Table.ExecuteAsync(TableOperation.Replace(entity)).ConfigureAwait(false);
-                }
-                else
+                throw new ArgumentException("Bogus etag in items with key: " + string.Join(", ", bogusEtagKeys.Select(o => o.Key)));
+            }
+
+            var table = await GetTable().ConfigureAwait(false);
+
+            var writeTasks = changes.Select(kv =>
+            {
+                var envelope = new StorageEnvelope(new EntityKey(kv.Key), kv.Value);
+
+                // Re-create object as table entity
+                var tableEntity = envelope.AsTableEntity();
+
+                if (envelope.ETag == null || envelope.ETag == "*")
                 {
-                    throw new Exception("etag empty");
+                    // New item or etag=* then insert or replace unconditionaly
+                    return table.ExecuteAsync(TableOperation.InsertOrReplace(tableEntity));
                 }
-            }
+
+
+                // Optimistic Update
+                return table.ExecuteAsync(TableOperation.Replace(tableEntity));
+            });
+
+            await Task.WhenAll(writeTasks).ConfigureAwait(false);
         }
 
-        protected class StoreItemEntity : TableEntity
+        /// <summary>
+        /// Ensure table is created.
+        /// </summary>
+        private ValueTask<CloudTable> GetTable()
         {
-            private static JsonSerializerSettings serializationSettings = new JsonSerializerSettings()
+            if (_table != null)
             {
-                // we use all so that we get typed roundtrip out of storage, but we don't use validation because we don't know what types are valid
-                TypeNameHandling = TypeNameHandling.All
-            };
+                return new ValueTask<CloudTable>(_table);
+            }
 
-            public StoreItemEntity() { }
+            return new ValueTask<CloudTable>(EnsureTableExists());
 
-            public StoreItemEntity(EntityKey key, object obj)
-                : this(key.PartitionKey, key.RowKey, obj)
-            { }
+            async Task<CloudTable> EnsureTableExists()
+            {
+                _table = _storageAccount.CreateCloudTableClient().GetTableReference(_tableName);
+                
+                // This call may not be thread-safe and multiple calls may be made, but Azure will handle it correctly and there is no destructive side effect.
+                await _table.CreateIfNotExistsAsync();
+                return _table;
```

Comment:
So this is why I used a local variable in my example. That way if two threads race to initialize, they are working with their own instances of `CloudTable` and the last one just "wins" in terms of publishing the local variable reference into the field.